api/service/stagedsync/stagedsync.go

```diff
@@ -0,0 +1,1321 @@
+		VerifyAllSig:           verifyAllSig,
+		VerifyHeaderBatchSize:  verifyHeaderBatchSize,
+		InsertChainBatchSize:   insertChainBatchSize,
+		LogProgress:            logProgress,
+	}
+}
+
+func (s *StagedSync) StageState(stage SyncStageID, tx kv.Tx, db kv.RwDB) (*StageState, error) {
+	var blockNum uint64
+	var err error
+	if errV := CreateView(context.Background(), db, tx, func(rtx kv.Tx) error {
+		blockNum, err = GetStageProgress(rtx, stage, s.isBeacon)
+		if err != nil {
+			return err
+		}
+		return nil
+	}); errV != nil {
+		return nil, errV
+	}
+
+	return &StageState{s, stage, blockNum}, nil
+}
+
+func (s *StagedSync) cleanUp(fromStage int, db kv.RwDB, tx kv.RwTx, firstCycle bool) error {
+	found := false
+	for i := 0; i < len(s.pruningOrder); i++ {
+		if s.pruningOrder[i].ID == s.stages[fromStage].ID {
+			found = true
+		}
+		if !found || s.pruningOrder[i] == nil || s.pruningOrder[i].Disabled {
+			continue
+		}
+		if err := s.pruneStage(firstCycle, s.pruningOrder[i], db, tx); err != nil {
+			panic(err)
+		}
+	}
+	return nil
+}
+
+func (s *StagedSync) Run(db kv.RwDB, tx kv.RwTx, firstCycle bool) error {
+	s.prevRevertPoint = nil
+	s.timings = s.timings[:0]
+
+	for !s.IsDone() {
+		var invalidBlockRevert bool
+		if s.revertPoint != nil {
+			for j := 0; j < len(s.revertOrder); j++ {
+				if s.revertOrder[j] == nil || s.revertOrder[j].Disabled {
+					continue
+				}
+				if err := s.revertStage(firstCycle, s.revertOrder[j], db, tx); err != nil {
+					return err
+				}
+			}
+			s.prevRevertPoint = s.revertPoint
+			s.revertPoint = nil
+			if s.invalidBlock != (common.Hash{}) {
+				invalidBlockRevert = true
+			}
+			s.invalidBlock = common.Hash{}
+			if err := s.SetCurrentStage(s.stages[0].ID); err != nil {
+				return err
+			}
+			firstCycle = false
+		}
+
+		stage := s.stages[s.currentStage]
+
+		if stage.Disabled {
+			utils.Logger().Trace().
+				Msgf("[STAGED_SYNC] %s disabled. %s", stage.ID, stage.DisabledDescription)
+
+			s.NextStage()
+			continue
+		}
+
+		if err := s.runStage(stage, db, tx, firstCycle, invalidBlockRevert); err != nil {
+			return err
+		}
+
+		s.NextStage()
+	}
+
+	if err := s.cleanUp(0, db, tx, firstCycle); err != nil {
+		return err
+	}
+	if err := s.SetCurrentStage(s.stages[0].ID); err != nil {
+		return err
+	}
+	if err := printLogs(tx, s.timings); err != nil {
+		return err
+	}
+	s.currentStage = 0
+	return nil
+}
+
+func CreateView(ctx context.Context, db kv.RwDB, tx kv.Tx, f func(tx kv.Tx) error) error {
+	if tx != nil {
+		return f(tx)
+	}
+	return db.View(context.Background(), func(etx kv.Tx) error {
+		return f(etx)
+	})
+}
+
+func ByteCount(b uint64) string {
+	const unit = 1024
+	if b < unit {
+		return fmt.Sprintf("%dB", b)
+	}
+	div, exp := uint64(unit), 0
+	for n := b / unit; n >= unit; n /= unit {
+		div *= unit
+		exp++
+	}
+	return fmt.Sprintf("%.1f%cB",
+		float64(b)/float64(div), "KMGTPE"[exp])
+}
+
+func printLogs(tx kv.RwTx, timings []Timing) error {
+	var logCtx []interface{}
```

Comment:
seems like this should be []string

Reply:
It is interface to allow to log everything. Current usage is only strings but can handle everything.